src/python/pants/backend/go/pkg.py

```diff
@@ -0,0 +1,256 @@
+from pants.engine.platform import Platform
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import UnexpandedTargets
+from pants.util.logging import LogLevel
+
+_logger = logging.getLogger(__name__)
+
+
+# A fully-resolved Go package. The metadata is obtained by invoking `go list` on the package.
+@dataclass(frozen=True)
+class ResolvedGoPackage:
+    # Address of the `go_package` target. Copied from the `ResolveGoPackageRequest` for ease of access.
+    address: Address
+
+    # Import path of this package. The import path will be inferred from an owning `go_module` if present.
+    import_path: str
+
+    # Address of the owning `go_module` if present. The owning `go_module` is the nearest go_module at the same
+    # or higher level of the source tree.
+    module_address: Optional[Address]
+
+    # Name of the package as given by `package` directives in the source files. Obtained from `Name` key in
+    # package metadata.
+    package_name: str
+
+    # Import paths used by this package. Obtained from `Imports` key in package metadata.
+    imports: Tuple[str, ...]
+
+    # Imports from test files. Obtained from `TestImports` key in package metadata.
+    test_imports: Tuple[str, ...]
+
+    # Explicit and transitive import paths required to build the code. Obtained from `Deps` key in package metadata.
+    dependency_import_paths: Tuple[str, ...]
+
+    # .go source files (excluding CgoFiles, TestGoFiles, XTestGoFiles). Obtained from `GoFiles` key in package metadata.
+    go_files: Tuple[str, ...]
+
+    # .go source files that import "C". Obtained from `CgoFiles` key in package metadata.
+    cgo_files: Tuple[str, ...]
+
+    # .go source files ignored due to build constraints. Obtained from `IgnoredGoFiles` key in package metadata.
+    ignored_go_files: Tuple[str, ...]
+
+    # non-.go source files ignored due to build constraints. Obtained from `IgnoredOtherFiles` key in package metadata.
+    ignored_other_files: Tuple[str, ...]
+
+    # _test.go files in package. Obtained from `TestGoFiles` key in package metadata.
+    test_go_files: Tuple[str, ...]
+
+    # _test.go files outside package. Obtained from `XTestGoFiles` key in package metadata.
+    xtest_go_files: Tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class ResolveGoPackageRequest:
+    address: Address
+
+
+def error_to_string(d: dict) -> str:
+    pos = d.get("Pos", "")
+    if pos:
+        pos = f"{pos}: "
+
+    import_stack_items = d.get("ImportStack", [])
+    import_stack = f" (import stack: {', '.join(import_stack_items)})" if import_stack_items else ""
+    return f"{pos}{d['Err']}{import_stack}"
+
+
+@rule
+async def resolve_go_package(
+    request: ResolveGoPackageRequest,
+    goroot: GoLangDistribution,
+    platform: Platform,
+    bash: BashBinary,
+) -> ResolvedGoPackage:
+    # TODO: Use MultiGet where applicable.
```

Comment:
(Not blocking) Would be good to use the format `TODO(#xxxx)` so that you can grep for this more easily. 

You do _not_ need a dedicated issue for this, of course: I'm suggesting a "catch-all" issue for the entire Go project.

Reply:
Will use this form after the cleanup phase next week. For now, I can just search for all TODO's in the backend packages.